### ISSUE TITLE
[Request] Galaxy Buds Live support

### ISSUE THREAD

**Issue by GioXmen** [ThePBone/GalaxyBudsClient#30]
Just putting this out for possible Galaxy Buds live support. This app made using the original buds with Windows 10 a joy due to the quicker pairing/connection.

I'm wondering if support could be added for Buds Live, especially since it seems Samsung hasn't even provided an official app for Windows this time.

Also finding out about hidden/locked away features is fun too 💯 

P.S: I do understand that for this to happen, you probably need to have the Buds Live on-hand to trace the Bluetooth communication messages.

**Comment by ThePBone** [comment]
Currently, I don't own the Galaxy Buds Live and I don't plan to buy a pair, mostly because they are lacking passive noise canceling and don't seal the ears completely off.
So there's no support for them planned in the near future (at least not from me). If anyone else has a good understanding of C# and knows a little bit about the Bluetooth infrastructure, then you're welcome to help to add support for the Buds Live by submitting a pull request (I could take care of the UI part in that case).

But maybe I'll also implement (basic) support for the Buds Live myself if no one else wants to do it. 
After all, I can refer to the disassembly of the Android plugin and 'evaluate' the raw Bluetooth messages by looking at (pseudo-)sources instead. (At first glance, the custom protocol used by the Buds Live doesn't look too different from the Buds+ one, so it shouldn't be too hard).
If I were to do it that way, I'd definitely need someone who would test my changes though, but I don't know when I'll find some time for this.

Anyway, I'll leave this open if anyone else is interested.

**Comment by GioXmen** [comment]
Understood. I'm also not the "biggest" fan of the new Buds Live, but I've received them with my Note 20 Ultra pre-order, as did many other people.

Between them and the OG buds these are marginally better (slightly lower sound quality, but much better connection/build and software quality).

I'll be keeping an eye on this, hopefully something will come out of this :) 

P.S: Perhaps a donation fundraiser would bring in enough funds to get you a pair for development purposes. I'd add a couple of bucks to that for sure.

**Comment by ThePBone** [comment]
Nevermind, I already added Buds Live support for it. All changes are mostly untested but it should work fine (hopefully).
I'm not sure if I implemented the touchpad long-press actions correctly though, they may be mixed up.

The updated source is available on a separate [neobeans](https://github.com/ThePBone/GalaxyBudsClient/commits/neobeans) (= Buds Live codename) branch right now.

Here's also a compiled version:
[GalaxyBudsClient_BudsLive_2.4-dev1.zip](https://github.com/ThePBone/GalaxyBudsClient/files/5105618/GalaxyBudsClient_BudsLive_2.4-dev1.zip)
Make sure you close all other instances of this app before launching the new executable (otherwise it'd just switch to the older one which was already active in the background).

Also to avoid confusion, this build opens another window on launch to inspect the raw incoming serial data (can be useful in some cases, you can dump all the received binary data into a file for example). 


It'd be cool if you could try it out and check if it is usable at all :)

**Comment by GioXmen** [comment]
Alright, I'll try it out! Thank you 👍 

I could also send the raw data dump if that helps you in any way.

**Comment by ThePBone** [comment]
Yes, that would be nice as well

**Comment by GioXmen** [comment]
I'm not sure which functionality has been restored/incorporated, so I'm just going to go from start to finish. (Haven't went through the code changes yet).

Before starting to test the new version, I did an uninstall of the old version, and also made sure the Program Files\Galaxy Buds Manager directory was empty.

- Front data panel:
![image](https://user-images.githubusercontent.com/6199566/90862410-10856a80-e396-11ea-806d-9a2c6b7adaa7.png)

- Seems to disconnect really easily, once that happens then it won't let me reconnect. Then I just have to de-register device and start from the beginning (Disconnection always happens when selecting a setting section and then going back to main menu). 
![image](https://user-images.githubusercontent.com/6199566/90862795-a4573680-e396-11ea-843c-9d8b5d64677f.png)

- Getting "too old firmware" notification on some settings - probably normal, as the application most likely does not recognize this new firmware version number.
![image](https://user-images.githubusercontent.com/6199566/90862769-9a353800-e396-11ea-9f27-d926a96299cd.png)

- No info in system information panel and self test:
![image](https://user-images.githubusercontent.com/6199566/90862897-ce105d80-e396-11ea-96c8-3d4d2a5c9596.png)

- The dev tool reported no incoming data, and the dump is empty (checked with text editor, its just an empty file).

So in summary, there must be issues in the basic connection, as it seems that it may not be getting past some initial startup connection? (I've tested to see if they are actually connected to Windows, and they were, music went through fine and the regular Stereo + Hands-free options could be seen, so its probably not the Buds themselves.)

**Comment by ManmohanGill** [comment]
seems like I'm getting this 

Seems to disconnect really easily, once that happens then it won't let me reconnect. Then I just have to de-register device and start from the beginning (Disconnection always happens when selecting a setting section and then going back to main menu).

![image](https://user-images.githubusercontent.com/18358271/91926704-66e59800-ed0a-11ea-9e01-edcf6fd948c5.png)

let me know how I get help to get the data as required 👍

**Comment by CPLT-C1** [comment]
I've also got a pair of Live's if you need me to do some testing - just let me know!

**Comment by followSol** [comment]
> seems like I'm getting this
> 
> Seems to disconnect really easily, once that happens then it won't let me reconnect. Then I just have to de-register device and start from the beginning (Disconnection always happens when selecting a setting section and then going back to main menu).
> 
> ![image](https://user-images.githubusercontent.com/18358271/91926704-66e59800-ed0a-11ea-9e01-edcf6fd948c5.png)
> 
> let me know how I get help to get the data as required 👍

I am having the same issue- I have one to test, Happy to help in any way and thanks, this looks super cool!

**Comment by ThePBone** [comment]
I haven't been able to continue working on this app lately but I'll try to explain what is going on here:

> So in summary, there must be issues in the basic connection, as it seems that it may not be getting past some initial startup connection? (I've tested to see if they are actually connected to Windows, and they were, music went through fine and the regular Stereo + Hands-free options could be seen, so its probably not the Buds themselves.)

Apparently, when registering a Galaxy Buds Live pair, the app is unable to connect to the serial protocol of the Buds at all, however, the app does not recognize this connection issue and still guides the user to the dashboard screen (with no headset connected). Therefore the app and the Buds did not exchange *any* data.
**All the information you are seeing are actually placeholders** and not real data because the fields are not correctly initialized.

Since this is caused by connection issues, it is hard for me to debug this without owning a pair.
If anyone knows how to use C# then you're welcome to send a pull request with potential fixes.
I'll also check my code a second time, as soon as I have set up a development environment for this in a VM.

**Comment by AndriesK** [comment]
I am not too familiar with C#, but it looks like it is using the Buds Service UUID, not the Buds Live's (same as Buds Plus), since the RegisteredDeviceModel does not have an option for Model.BudsLive, with this option added, it seems to work fine, looks like just a small oversight? See: #40

**Comment by ThePBone** [comment]
@AndriesK Thanks a lot! I can't believe I made this dumb mistake. :sweat_smile: 
I've merged it but I'm not near my development machine for a few days, so I can't prepare a (pre-)release for this.

Did you experience any other bugs with the Buds Live? Are the settings/switches specific to the Buds Live properly working?

**Comment by JojiiOfficial** [comment]
I have buds live. If i was able to compile it, I'd test it but as I was told, this seems to use legacy dependencies which makes it even harder to compile it for me. Are there any instructions for a non C# .Net Framework dev?

**Comment by AndriesK** [comment]
@ThePBone Haha it happens, I'm just happy to be able to use this! No issues/bugs for me, looks like all the switches/settings are working great, but I'll comment here if I find something odd.

**Comment by ThePBone** [comment]
By the way, you don't need to install the UWP module to compile this project. It is only required to build signed Windows Store packages but works just fine if you tell Visual Studio to skip installing these dependencies when asked.

**Comment by JojiiOfficial** [comment]
@AndriesK I didn't have VisualStudio installed. I tried to compile it without vs but this (kinda obv. tbh) didn't work out that well. After installing vs later on I got it to work. It works perfectly using my Buds live!

**Comment by ThePBone** [comment]
Can someone test this debug build? It introduces Buds Live support but is not yet finished. I need some more information about what exact data the Buds Live transmit, in order to make sure I didn't overlook something.

Also, can anyone who owns a pair send me a memory dump of the incoming data?
(Like this: Go to 'File > Dump to file...' in the second debug window that opens in the background. Just make sure to wait a bit until some data was collected)

[Setup_GalaxyBudsManager_2.3.3.5b.zip](https://github.com/ThePBone/GalaxyBudsClient/files/5371161/Setup_GalaxyBudsManager_2.3.3.5b.zip)

**Comment by ThePBone** [comment]
>If I touch and hold the buds, I can hear the sound that ANC turned on, but the UI does not reflect that state.

Looks like I forgot to listen for `MSG_ID_NOISE_REDUCTION_MODE_UPDATE` messages, this should hopefully fix it:

[Setup_GalaxyBudsManager_2.3.3.6b.zip](https://github.com/ThePBone/GalaxyBudsClient/files/5372312/Setup_GalaxyBudsManager_2.3.3.6b.zip)

Thanks for sending a dump, I'll have a look at it later.

**Comment by AndriesK** [comment]
Ah, cool it's working on that version. thanks

**Comment by ThePBone** [comment]
Great! I haven't found anything out of the ordinary in the dump either, so I can start preparing a new release. You can probably expect it within two days (if there are no other issues found).

**Comment by followSol** [comment]
For the record I’ve been using the Buds Live with the latest version you shared and so far so good. Thanks again!

**Comment by ThePBone** [comment]
Version 3.0 introduces Galaxy Buds Live support.
-> https://github.com/ThePBone/GalaxyBudsClient/releases/tag/3.0.0.0

Thank you to everyone who helped to make this work!
